cabal-install/src/Distribution/Client/SetupWrapper.hs

```diff
@@ -851,285 +988,409 @@ getExternalSetupMethod verbosity options pkg bt = do
-                      toNubListR
-                        [ cppMacrosFile
-                        | useVersionMacros options'
-                        ]
-                  , ghcOptExtra = extraOpts
-                  }
-          let ghcCmdLine = renderGhcOptions compiler platform ghcOptions
-          when (useVersionMacros options') $
-            rewriteFileEx verbosity (i cppMacrosFile) $
-              generatePackageVersionMacros (pkgVersion $ package pkg) (map snd selectedDeps)
-          case useLoggingHandle options of
-            Nothing -> runDbProgramCwd verbosity mbWorkDir program progdb ghcCmdLine
-            -- If build logging is enabled, redirect compiler output to
-            -- the log file.
-            Just logHandle -> do
-              output <-
-                getDbProgramOutputCwd
-                  verbosity
-                  mbWorkDir
-                  program
-                  progdb
-                  ghcCmdLine
-              hPutStr logHandle output
-        return $ i setupProgFile
-
-isCabalPkgId :: PackageIdentifier -> Bool
+          else do
+            debug verbosity $ "Setup executable not found in the cache."
+            src <-
+              compileSetupExecutable
+                verbosity
+                platform
+                pkgId
+                bt
+                options'
+                cabalLibVersion
+                maybeCabalLibInstalledPkgId
+                True
+            createDirectoryIfMissingVerbose verbosity True setupCacheDir
+            installExecutableFile verbosity src cachedSetupProgFile
+            -- Do not strip if we're using GHCJS, since the result may be a script
+            when (maybe True ((/= GHCJS) . compilerFlavor) $ useCompiler options') $ do
+              -- Add the relevant PATH overrides for the package to the
+              -- program database.
+              setupProgDb
+                <- prependProgramSearchPath verbosity
+                      (useExtraPathEnv options')
+                      (useExtraEnvOverrides options')
+                      (useProgramDb options')
+                     >>= configureAllKnownPrograms verbosity
+              Strip.stripExe
+                verbosity
+                platform
+                setupProgDb
+                cachedSetupProgFile
+    return cachedSetupProgFile
+    where
+      criticalSection' = maybe id criticalSection $ setupCacheLock options'
+
+-- | If the Setup.hs is out of date wrt the executable then recompile it.
+-- Currently this is GHC/GHCJS only. It should really be generalised.
+compileSetupExecutable
+  :: Verbosity
+  -> Platform
+  -> PackageIdentifier
+  -> BuildType
+  -> SetupScriptOptions
+  -> Version
+  -> Maybe ComponentId
+  -> Bool
+  -> IO FilePath
+compileCustomSetupExecutable, compileHooksExecutable
+  :: Verbosity
+  -> Platform
+  -> PackageIdentifier
+  -> SetupScriptOptions
+  -> Version
+  -> Maybe ComponentId
+  -> Bool
+  -> IO FilePath
```

Comment:
This is de gustibus but I find this kind of interleaving hard to read. This in particular feels very dense given the similarity in names (compileSetupExecutable, compileCustomSetupExecutable, compileHooksExecutable, compileSetupExecutableX).

Reply:
After playing with this part for a bit I got a bit confused. `compileSetupExecutable` can be called for any build-type, not only Custom or Hooks. Meanwhile here, for `bt /= Hook` we call `compileCustomSetupExecutable` which in turn calls `compileSetupExecutableX` with `Custom` as build-type.

Is this intended? Are we reusing BuildType to convey additional operational meaning? I am wondering if we need to more/better types.

Reply:
Yeah, I actually got confused for a while during refactoring because of exactly this point, so it's definitely an appropriate comment.

Reply:
I pushed some changes that should make things clearer. Let me know how you find it.